File: test-platform-v2/backend/app/core/execution_status.py
"""执行状态统一词表与规范化（FIX-173-P1-06 / Batch 182）。

历史：同一执行事实在 5 张表 4 套词表（pass/fail/skip/block、success、
done/fail、completed/failed、passed/failed/skipped），聚合层被迫逐表映射。
Batch 182 统一为单一规范值集合：

    pending | running | passed | failed | skipped | cancelled | blocked

- 新代码只写规范值；
- `canonical_exec_status(v)` 兼容历史旧值（pass/fail/skip/block/success/done/completed），
  供 open_api 回写（CI 契约向后兼容）与过渡期读取使用；
- 前端经 `frontend/src/utils/executionStatus.ts` 展示（新旧双值兼容）。
"""
from __future__ import annotations

# 规范状态值（DB 唯一词表）
PENDING = "pending"
RUNNING = "running"
PASSED = "passed"
FAILED = "failed"
SKIPPED = "skipped"
CANCELLED = "cancelled"
BLOCKED = "blocked"

CANONICAL_STATUSES = frozenset({PENDING, RUNNING, PASSED, FAILED, SKIPPED, CANCELLED, BLOCKED})

# 旧值 → 规范值（历史数据/外部契约兼容映射）
_LEGACY_MAP = {
    "pass": PASSED,
    "fail": FAILED,
    "skip": SKIPPED,
    "block": BLOCKED,
    "success": PASSED,
    "done": PASSED,
    "completed": PASSED,
}


def canonical_exec_status(value: str | None) -> str:
    """把任意历史/外部状态值规范化为统一词表；未知值原样返回。

    open_api 回写等外部入口必须经本函数后落库，保证 CI 旧脚本（pass/fail/skip/block）
    与新值并存期间行为一致。
    """
    if not value:
        return value or ""
    if value in CANONICAL_STATUSES:
        return value
    return _LEGACY_MAP.get(value, value)

File: test-platform-v2/frontend/src/utils/executionStatus.ts
/**
 * 执行状态统一映射（Batch 182 / FIX-173-P1-06）
 *
 * 后端 DB/API 已统一词表：pending | running | passed | failed | skipped | cancelled | blocked
 * 本模块提供：
 *  - EXEC_STATUS_LABEL：新旧双值 → 中文标签（历史数据/过渡期兼容展示）
 *  - normalizeExecStatus(v)：旧值归一为规范值（展示层用规范值查映射）
 */
export const EXEC_STATUS_LABEL: Record<string, string> = {
  // 规范值
  pending: '待执行',
  running: '执行中',
  passed: '通过',
  failed: '失败',
  skipped: '跳过',
  cancelled: '已取消',
  blocked: '阻塞',
  // 历史/过渡值（后端已迁移，兼容旧前端缓存与旧数据）
  pass: '通过',
  fail: '失败',
  skip: '跳过',
  block: '阻塞',
  success: '通过',
  done: '通过',
  completed: '通过',
  idle: '待执行',
}

/** 旧值 → 规范值（与后端 app/core/execution_status.py 保持一致） */
const LEGACY_TO_CANONICAL: Record<string, string> = {
  pass: 'passed',
  fail: 'failed',
  skip: 'skipped',
  block: 'blocked',
  success: 'passed',
  done: 'passed',
  completed: 'passed',
  idle: 'pending',
}

export function normalizeExecStatus(value?: string | null): string {
  if (!value) return value ?? ''
  if (LEGACY_TO_CANONICAL[value]) return LEGACY_TO_CANONICAL[value]
  return value
}

/** 展示标签：旧值/新值都可用 */
export function execStatusLabel(value?: string | null): string {
  return EXEC_STATUS_LABEL[value ?? ''] ?? value ?? ''
}
